### DISCUSSION TITLE
scheduleRecurrently doesn't work when the table has 19 million rows.

### DISCUSSION THREAD

**Discussion by rjhho** [jobrunr/jobrunr#1174] [Q&A]
Maven version: apache-maven-3.6.3
jdk: 1.8
springboot: 2.7.18
database: Oracle Database 12c

Hi jobrunr community,

I've implemented the schedulerecurrently method and so far it has been running smoothly, but after i imported 19 million rows of messages into my "Berichten" table, the schedulerecurrently doesn't seem to work anymore. 

The confusing part is that a copy of the exact same database with only 100 rows of messages in the "Berichten" table seem to work totally fine.

I commented out the methods that interact with the database assuming that that would make de schedulRecurrently run again, but no success.

Do you guys know what is going on here?

some more info: 
the setup is almost the same as my example project, 
https://github.com/rjhho/RecurrentlyTestProject/tree/main
the only difference is that i call an actual database to update my "Berichten" table. 

The actual database is called "Oracle Database 12c"

in this database i also added the jobrunr tables,indexes,view by following the steps 
(Setup a custom user only for the Database Migrations + Apply the SQL scripts yourself)
on this page:
https://www.jobrunr.io/en/documentation/pro/database-migrations/ 
the only hurdle was that i had to add "create any view" rights to add the "JOBRUNR_JOBS_STATS" VIEW.

I also implemented the commandlinerunner to run the scheduleRecurrently method below.

    @Override
    public void run(String... args) throws Exception {
        BackgroundJob.scheduleRecurrently("id_taak_1", "0 0 * * *",
            () -> berichtService.updateBericht(20001);
    }

I hope this is enough info to describe my issue 

Thanks in advance 

Raymond

**Comment by rdehuyss** [comment]
Hi @rjhho - I think you're bumping into the limit of JobRunr OSS where we've not optimized it for that many recurring jobs. In Pro all queries are optimized to handle much more load.

<img width="1145" alt="image" src="https://github.com/user-attachments/assets/7a62c680-ed17-47c5-93c9-c4defd25e543" />

**Reply by rjhho** [reply]
Hi @rdehuyss ,
currently I'm only using the scheduleRecurrently to run 1 recurring job once every night at 00:00 hour (cron = "0 0 * * *")

sry i made a typo in the cron string value, 
it is:
BackgroundJob.scheduleRecurrently("id_taak_1", "0 0 * * * *",
and not:
BackgroundJob.scheduleRecurrently("id_taak_1", "*/5 * * * * *",

To be more specific, my question is as follows is it possible to run the following scheduleJob in the "Berichten" table with the 19 million rows? I'm genuinely confused why this currently works in my database with 100 rows of messages, but that it doesn't work in a database with 19 million rows. 

```
@Override
public void run(String... args) throws Exception {
    BackgroundJob.scheduleRecurrently("id_taak_1", "0 0 * * *",  
        () -> berichtService.updateBericht(20001);
}
```
berichtservice method:
```
    public int updateBericht(int id) {
        System.out.println("jdbctemplate is: " + jdbcTemplate);
        System.out.println("code has been reached aka updateBericht");
        String sql = "UPDATE Berichten SET message='TEST2' WHERE id=" + id;
        return jdbcTemplate.update(sql);
    }
```

Commenting out the sql function also fails to make the method schedulerRecurrently work again and doesn't print out the system.out.println values, so i assume that schedulerRecurrently doesn't even try to enter the updateBericht method:
```
    public int updateBericht(int id) {
        System.out.println("jdbctemplate is: " + jdbcTemplate);
        System.out.println("code has been reached aka updateBericht");
        //String sql = "UPDATE Berichten SET message='TEST2' WHERE id=" + id;
        //return jdbcTemplate.update(sql);
        return 1;
    }
```
Am i overlooking something?

Thanks for your time

Raymond

**Reply by auloin** [reply]
Hi @rjhho, this doesn't sound like a JobRunr issue to me... Are you using the dashboard? It can allow you to see the state of your system (Jobs, RecurringJobs, and BackgroundJobServers included).

**Reply by rjhho** [reply]
Hi @auloin thanks for the input. 

I have a question about the dashboard: If I have 1 springboot project (projectA) that focuses on running the schedule recurrently
and another springboot project (projectB) that only exist to run the dashboard. 
Is it possible to see in the dashboard from projectB the state of the jobs from projectA as long as both projects have the same datasource.url? (So both will share the same JOBRUNR tables)

Thanks for your time

**Reply by auloin** [reply]
@rjhho yes, it's.

**Reply by rjhho** [reply]
hi @auloin, 
The dashboard doesn't seem to run, after setting the log level to DEBUG, it looks like JobRunrAutoConfiguration failed to do something with the dashboard. 

``` 
JobRunrAutoConfiguration#dashboardWebServer:
      Did not match:
         - @ConditionalOnProperty (org.jobrunr.dashboard.enabled=true) found different value in property 'enabled' (OnPropertyCondition)

JobRunrAutoConfiguration#dashboardWebServerConfiguration:
      Did not match:
         - @ConditionalOnProperty (org.jobrunr.dashboard.enabled=true) found different value in property 'enabled' (OnPropertyCondition)
``` 

These are the only dependencies that i used

``` 
     <dependencies>
        <dependency>
            <groupId>org.jobrunr</groupId>
            <artifactId>jobrunr-spring-boot-2-starter</artifactId>
            <version>7.2.0</version>
        </dependency>

        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-web</artifactId>
            <version>2.7.18</version>
        </dependency>
    </dependencies>
``` 
this is the only code that i use

``` 
@Configuration
public class MainConfiguration {
    @Bean
    public StorageProvider storageProvider(JobMapper jobMapper) {
        InMemoryStorageProvider storageProvider = new InMemoryStorageProvider();
        storageProvider.setJobMapper(jobMapper);
        return storageProvider;
    }
}
``` 
and 

``` 
@SpringBootApplication
@Import(MainConfiguration.class)
public class MainApplication {
    public static void main(String[] args) {
        SpringApplication.run(MainApplication.class, args);
    }
}
``` 
My application.properties looks like this:

``` 
org.jobrunr.background-job-server.enabled=false
org.jobrunr.dashboard.enabled=true 
org.jobrunr.dashboard.port=8080
logging.level.root=debug
``` 
Do you know what's going on here?

Thanks in advance!

**Reply by auloin** [reply]
@rjhho I've no idea, if you're able to find a fault in enabling the dashboard using any of the examples we provide: https://github.com/jobrunr/example-spring, https://github.com/jobrunr/example-java-mag or https://github.com/jobrunr/example-order-fulfillment, we can investigate. Otherwise we'll not be able to help as the issue is not from JobRunr.

**Reply by rjhho** [reply]
Hi @auloin, thanks for your quick reply. We managed to make the dashboard work just now.   Next step is to connect it to the large db 🙏 
to be continued
